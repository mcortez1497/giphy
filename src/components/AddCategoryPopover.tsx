import React, { ChangeEvent, KeyboardEvent } from "react";

import {
  Button,
  IconButton,
  InputAdornment,
  Paper,
  TextField,
  WithStyles
} from "@material-ui/core";
import { createStyles, Theme, withStyles } from "@material-ui/core/styles";
import { Add } from "@material-ui/icons";

const styles = (theme: Theme) =>
  createStyles({
    root: {
      padding: theme.spacing(0, 1, 1, 1)
    }
  });

interface Props extends WithStyles<typeof styles> {
  readonly onClose: () => void;
  readonly onSubmit: (categoryName: string) => void;
}

interface State {
  categoryName: string;
}

class AddCategoryPopoverWithStyles extends React.Component<Props, State> {
  state = {
    categoryName: ""
  };

  render() {
    const {
      handleCategoryNameChange,
      handleSubmit,
      handleSubmitOnKeypress,
      props: { classes },
      state: { categoryName }
    } = this;

    const isValid = categoryName !== "";

    return (
      <div className={classes.root}>
        <TextField
          variant="outlined"
          margin="normal"
          required
          fullWidth
          id="category"
          label="Category Name"
          name="category"
          value={categoryName}
          onChange={handleCategoryNameChange}
          onKeyPress={handleSubmitOnKeypress}
          autoFocus
          InputProps={{
            endAdornment: (
              <InputAdornment position="end">
                <IconButton
                  disabled={!isValid}
                  edge="end"
                  aria-label="add category"
                  onClick={handleSubmit}
                >
                  <Add color={isValid ? "primary" : "disabled"} />
                </IconButton>
              </InputAdornment>
            )
          }}
        />
      </div>
    );
  }

  private handleSubmit = () => {
    if (this.state.categoryName === "") {
      return;
    }
    this.props.onSubmit(this.state.categoryName);
    this.props.onClose();
  };

  private handleSubmitOnKeypress = (event: KeyboardEvent<HTMLDivElement>) => {
    if (event.key === "Enter") {
      this.handleSubmit();
    }
  };

  private handleCategoryNameChange = (event: ChangeEvent<HTMLInputElement>) => {
    this.setState({ categoryName: event.target.value });
  };
}

const AddCategoryPopover = withStyles(styles)(AddCategoryPopoverWithStyles);

export { AddCategoryPopover };
