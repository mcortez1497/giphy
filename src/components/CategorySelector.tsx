import React, { ChangeEvent } from "react";

import {
  Chip,
  FormControl,
  Input,
  InputLabel,
  MenuItem,
  Select,
  WithStyles
} from "@material-ui/core";
import { createStyles, Theme, withStyles } from "@material-ui/core/styles";

import { Category, Gif } from "types";

const styles = (theme: Theme) =>
  createStyles({
    chips: {
      display: "flex",
      flexWrap: "wrap"
    },
    chip: {
      margin: 2
    },
    grow: {
      flexGrow: 1
    },
    label: {
      top: "-7px"
    },
    select: {
      flexGrow: 1,
      "&::before": {
        borderWidth: 0
      }
    }
  });

const MenuProps = {
  PaperProps: {
    style: {
      maxHeight: 48 * 4.5 + 8,
      width: 250
    }
  }
};

interface Props extends WithStyles<typeof styles> {
  readonly categories: Category[];
  readonly gif: Gif;

  readonly onCategorySelect: (gifId: string, categories: string[]) => void;
}

const CategorySelectorWithStyles: React.FC<Props> = ({
  categories,
  classes,
  gif,
  onCategorySelect
}) => (
  <FormControl className={classes.grow}>
    {!gif.categories ||
      (gif.categories.length <= 0 && (
        <InputLabel htmlFor="select-multiple-chip" className={classes.label}>
          Add Categories
        </InputLabel>
      ))}
    <Select
      multiple
      value={getSelectedCategories(gif)}
      onChange={handleCategorySelect(gif, onCategorySelect)}
      className={classes.select}
      input={<Input id="select-multiple-chip" />}
      renderValue={selected => (
        <div className={classes.chips}>
          {(selected as string[]).map(categoryId => (
            <Chip
              key={categoryId}
              label={getCategoryNameById(categoryId, gif.categories)}
              className={classes.chip}
            />
          ))}
        </div>
      )}
      MenuProps={MenuProps}
      IconComponent={() => null}
    >
      >
      <MenuItem value="" disabled>
        Add Categories
      </MenuItem>
      {categories.map((category, index) => (
        <MenuItem
          key={index}
          value={category._id}
        >
          {category.name}
        </MenuItem>
      ))}
    </Select>
  </FormControl>
);

const getSelectedCategories = (gif: Gif) =>
  gif.categories ? gif.categories.map(cat => cat._id) : [];

const getCategoryNameById = (
  categoryId: string,
  categories: Category[] = []
) => {
  const category = categories.find(category => category._id === categoryId);
  return category ? category.name : "";
};

const handleCategorySelect = (gif: Gif, onCategorySelect: Function) => (
  event: ChangeEvent<{ name?: string | undefined; value: any }>
) => {
  if (!gif._id) {
    return;
  }
  onCategorySelect(gif._id, event.target.value);
};

const CategorySelector = withStyles(styles)(CategorySelectorWithStyles);

export { CategorySelector };
