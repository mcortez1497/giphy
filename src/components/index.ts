export * from './Content';
export * from './ContentContainer';
export * from "./GifCard";
export * from "./GifCardContainer";
export * from "./Header";
export * from './HeaderContainer';
export * from "./LoginPopover";
export * from "./RegistrationPopover";
export * from "./SearchBar";
export * from "./UserDrawer";
export * from "./UserDrawerContainer";