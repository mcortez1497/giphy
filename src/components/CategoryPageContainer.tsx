import * as React from "react";
import { connect } from "react-redux";
import { RouteComponentProps } from "react-router";
import { bindActionCreators, Dispatch } from "redux";

import { Content, Layout } from "components";
import { AppState, getGifs } from "reducers";
import { Gif } from "types";

interface StateProps {
  readonly gifs: Gif[];
}

interface DispatchProps {
  readonly getGifs: () => void;
}

interface RouteProps {
  readonly categoryId: string;
}

interface ComponentProps
  extends StateProps,
    DispatchProps,
    RouteComponentProps<RouteProps> {}

class Container extends React.Component<ComponentProps> {
  public componentDidMount() {
    this.props.getGifs();
  }

  public render() {
    return (
      <Layout>
        <Content gifs={this.props.gifs} />
      </Layout>
    );
  }
}

const mapStateToProps = (
  state: AppState,
  props: RouteComponentProps<RouteProps>
) => ({
  gifs: state.user.gifs.reduce(
    (gifsWithCategory: Gif[], gif: Gif) =>
      gif.categories &&
      gif.categories.find(
        category => category._id === props.match.params.categoryId
      )
        ? [...gifsWithCategory, gif]
        : gifsWithCategory,
    []
  )
});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      getGifs
    },
    dispatch
  );

export const CategoryPageContainer = connect<
  StateProps,
  DispatchProps,
  RouteComponentProps<RouteProps>,
  AppState
>(
  mapStateToProps,
  mapDispatchToProps
)(Container);
