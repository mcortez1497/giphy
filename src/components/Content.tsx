import React from "react";
import { GifViewer } from "../components";
import { Gif } from "../types";

interface Props {}

interface State {
  gifs: Gif[];
}

class Content extends React.Component<Props, State> {
  state = {
    gifs: []
  };

  async componentDidMount() {
    fetch(
      "http://api.giphy.com/v1/gifs/trending?api_key=osVl8xeSb13OvfguXaI0pUrbRzph61v9"
    )
      .then(response => response.json())
      .then(json => {
        const gifs = json.data.map((gif: Gif) => gif);
        this.setState({
          gifs
        });
      })
      .catch(error => console.log(error));
  }

  render() {
    return (
      <div>
        <div>
          <input type="text" placeholder="Search for GIFs" />
        </div>
        <GifViewer gifs={this.state.gifs} />
      </div>
    );
  }
}

export { Content };
