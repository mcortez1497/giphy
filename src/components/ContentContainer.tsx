import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';

import { AppState, getGifs } from 'reducers';
import { Gif } from 'types';

import { Content } from 'components';

interface StateProps {
  readonly gifs: Gif[];
}

interface DispatchProps {
  readonly getGifs: () => void;
}

interface ComponentProps extends StateProps, DispatchProps {}

class Container extends React.Component<ComponentProps> {
  public componentDidMount() {
    this.props.getGifs();
  }

  public render() {
    return <Content gifs={this.props.gifs} />;
  }
}

const mapStateToProps = (state: AppState) => ({
  gifs: state.gifs.items
})

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      getGifs
    },
    dispatch
  );

export const ContentContainer = connect<StateProps, DispatchProps, {}, AppState>(
  mapStateToProps,
  mapDispatchToProps
)(Container);
