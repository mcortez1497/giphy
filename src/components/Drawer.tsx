import React from "react";
import { Link } from "react-router-dom";

import {
  Divider,
  Drawer as TemporaryDrawer,
  IconButton,
  List,
  ListItem,
  ListItemIcon,
  ListItemText,
  Popover,
  WithStyles,
  Typography
} from "@material-ui/core";
import { createStyles, Theme, withStyles } from "@material-ui/core/styles";
import {
  Add,
  Category as CategoryIcon,
  Favorite,
  Home
} from "@material-ui/icons";

import { AddCategoryPopover } from "components";
import { Category } from "types";

const styles = (theme: Theme) =>
  createStyles({
    addCategory: {
      padding: theme.spacing(0.5)
    },
    drawer: {
      width: 250
    },
    nested: {
      paddingLeft: theme.spacing(4)
    },
    title: {
      display: "flex",
      flexDirection: "column",
      alignItems: "flex-start",
      padding: theme.spacing(2)
    }
  });

interface Props extends WithStyles<typeof styles> {
  categories: Category[];
  isOpen: boolean;

  addCategory: (categroyName: string) => void;
  closeDrawer: () => void;
}

interface State {
  anchorEl: HTMLButtonElement | null;
}

class DrawerWithStyles extends React.Component<Props, State> {
  state = {
    anchorEl: null
  };

  render() {
    const {
      handlePopoverClose,
      handlePopoverOpen,
      props: { addCategory, categories, classes, closeDrawer, isOpen = false },
      state: { anchorEl }
    } = this;

    const open = Boolean(anchorEl);
    const id = open ? "add-category-popover" : undefined;

    return (
      <TemporaryDrawer open={isOpen} onClose={closeDrawer}>
        <div className={classes.drawer}>
          <ListItem className={classes.title}>
            <Typography variant="h5" color="primary">
              GIF Viewer
            </Typography>
            <Typography variant="caption">
              By Michael Cortez for H-E-B
            </Typography>
          </ListItem>
          <Divider />
          <ListItem
            button
            component={Link}
            to="/"
            onClick={closeDrawer}>
            <ListItemIcon>
              <Home />
            </ListItemIcon>
            <ListItemText primary={"Home"} />
          </ListItem>
          <ListItem
            button
            component={Link}
            to="/my-gifs"
            onClick={closeDrawer}
          >
            <ListItemIcon>
              <Favorite />
            </ListItemIcon>
            <ListItemText primary={"My Saved Gifs"} />
          </ListItem>
          <ListItem>
            <ListItemIcon>
              <CategoryIcon />
            </ListItemIcon>
            <ListItemText primary={"Categories"} />
            <IconButton
              aria-describedby={id}
              aria-label="add category"
              className={classes.addCategory}
              onClick={handlePopoverOpen}
            >
              <Add />
            </IconButton>
            <Popover
              id={id}
              open={open}
              anchorEl={anchorEl}
              onClose={handlePopoverClose}
              anchorOrigin={{
                vertical: "center",
                horizontal: "right"
              }}
              transformOrigin={{
                vertical: "center",
                horizontal: "left"
              }}
            >
              <AddCategoryPopover
                onSubmit={addCategory}
                onClose={handlePopoverClose}
              />
            </Popover>
          </ListItem>
          <List component="div" disablePadding>
            {categories.map((category, index) => (
              <ListItem
                button
                key={index}
                className={classes.nested}
                component={Link}
                to={`/categories/${category._id}`}
                onClick={closeDrawer}
              >
                <ListItemText primary={category.name} />
              </ListItem>
            ))}
          </List>
        </div>
      </TemporaryDrawer>
    );
  }

  private handlePopoverOpen = (
    event: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) =>
    this.setState({
      anchorEl: event.currentTarget
    });

  private handlePopoverClose = () =>
    this.setState({
      anchorEl: null
    });
}

const Drawer = withStyles(styles)(DrawerWithStyles);

export { Drawer };
