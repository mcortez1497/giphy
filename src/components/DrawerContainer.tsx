import { connect } from "react-redux";
import { bindActionCreators, Dispatch } from "redux";

import { Drawer } from "components"
import { AppState, addCategory, setDrawerClosed } from "reducers";
import { Category } from "types";

interface StateProps {
  readonly categories: Category[];
  readonly isOpen: boolean;
}

interface DispatchProps {
  readonly addCategory: (categoryName: string) => void;
  readonly closeDrawer: () => void;
}

const mapStateToProps = (state: AppState) => ({
  categories: state.user.categories.items.sort((a, b) => a.name > b.name ? 1 : -1),
  isOpen: state.ui.isDrawerOpen
});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      addCategory,
      closeDrawer: setDrawerClosed
    },
    dispatch
  );

export const DrawerContainer = connect<
  StateProps,
  DispatchProps,
  {},
  AppState
>(
  mapStateToProps,
  mapDispatchToProps
)(Drawer);
