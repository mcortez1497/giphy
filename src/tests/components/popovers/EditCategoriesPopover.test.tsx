import { EditCategoriesPopover } from "components";
import { TestHelper } from "tests";

const props = {
  hasCategories: true,
  onClose: jest.fn(),
  onAddClick: jest.fn(),
  ronRemoveClick: jest.fn()
};

describe("EditCategoriesPopover", () => {
  describe("Rendering", () => {
    it("renders correctly", () => {
      const { element } = TestHelper.getInstanceObjects(
        EditCategoriesPopover,
        props
      );
      expect(element).toMatchSnapshot();
    });

    it("renders correctly if there are no categories", () => {
      const { element } = TestHelper.getInstanceObjects(EditCategoriesPopover, {
        ...props,
        hasCategories: false
      });
      expect(element).toMatchSnapshot();
    });
  });
});
