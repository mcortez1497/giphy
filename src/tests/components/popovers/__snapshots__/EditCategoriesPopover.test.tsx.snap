// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditCategoriesPopover Rendering renders correctly 1`] = `
<ForwardRef(MenuList)>
  <ForwardRef(WithStyles)
    onClick={[Function]}
  >
    Add
  </ForwardRef(WithStyles)>
  <ForwardRef(WithStyles)
    disabled={false}
    onClick={[Function]}
  >
    Remove...
  </ForwardRef(WithStyles)>
</ForwardRef(MenuList)>
`;

exports[`EditCategoriesPopover Rendering renders correctly if there are no categories 1`] = `
<ForwardRef(MenuList)>
  <ForwardRef(WithStyles)
    onClick={[Function]}
  >
    Add
  </ForwardRef(WithStyles)>
  <ForwardRef(WithStyles)
    disabled={true}
    onClick={[Function]}
  >
    Remove...
  </ForwardRef(WithStyles)>
</ForwardRef(MenuList)>
`;
